async_ssh2_russh/src/promise.rs

```diff
@@ -0,0 +1,195 @@
+    /// was dropped before sending a value.
+    pub fn try_get(&self) -> Result<&T, PromiseError> {
+        self.inner
+            .get()
+            .ok_or(PromiseError::Empty)
+            .and_then(|value_opt| value_opt.ok_or(PromiseError::Dropped))
+    }
+}
+
+/// An error that may occur when trying to get the value from a [`Promise`], see [`Promise::try_get`].
+#[derive(Debug, Eq, PartialEq, Clone, thiserror::Error)]
+pub enum PromiseError {
+    /// The resolver has not yet sent a value.
+    #[error("value not yet sent")]
+    Empty,
+    /// The resolver was dropped before sending a value.
+    #[error("closed before a value was sent")]
+    Dropped,
+}
+
+// SAFETY: must not be clonable, as that would allow simultaneous write access to the `Inner` value.
+/// The resolve/send half of a promise. Created via [`channel`], and used to resolve the corresponding [`Promise`] with
+/// a value.
+pub struct Resolve<T> {
+    inner: Option<Weak<Inner<T>>>,
+}
+impl<T> Resolve<T> {
+    /// Resolves the promise with the given value, consuming this `Resolve` in the process.
+    ///
+    /// This will panic if the promise has already been resolved.
+    pub fn into_resolve(mut self, value: T) {
+        self.resolve(value).unwrap_or_else(|_| panic!("already resolved"));
+    }
+
+    /// Resolves the promise with the given value, returning an error if the promise has already been resolved.
+    pub fn resolve(&mut self, value: T) -> Result<(), T> {
+        let Some(inner) = self.inner.take() else {
+            return Err(value);
+        };
+
+        if let Some(inner) = inner.upgrade() {
+            // SAFETY: `&mut self: Resolve` has exclusive access to `resolve` once.
+            unsafe {
+                inner.resolve(Some(value));
+            }
+        }
+        Ok(())
+    }
+}
+impl<T> Drop for Resolve<T> {
+    fn drop(&mut self) {
+        if let Some(inner) = self.inner.take().and_then(|weak| weak.upgrade()) {
+            // SAFETY: `&mut self: Resolve` has exclusive access to call resolve once. Because we use `inner.take()`, we
+            // know `Self::resolve` was not called.
+            unsafe {
+                inner.resolve(None);
+            }
+        }
+    }
+}
+
+/// Flag for when the [`Resolve`] will no longer change the value of the [`Inner`], because it was either resolved or
+/// dropped.
+const FLAG_COMPLETED: u8 = 0b01;
+/// Flag for when the value is set and can be read.
+const FLAG_VALUE_SET: u8 = 0b10;
+
+/// Any thread with an `&self` may access the `value` field according the following rules:
+///
+///  1. Iff `flag & FLAG_COMPLETED` is false, the `value` field may be modified by the owning `Resolve` once.
+///  2. Iff `flag & FLAG_VALUE_SET` is true, the `value` field may be accessed immutably by any thread.
+///
+/// If `flag & FLAG_COMPLETED` is true, but `flag & FLAG_VALUE_SET` is false, then the owning `Resolve` was dropped
```

Comment:
It's a bit of extra work to read this comment. There are 4 settings of the bits. A little table of all 4 cases with a comment for each would be easier to read.

Reply:
Also while I'm all for documenting this, it seems like it's not proximate to where the cases are handled, which is a bit annoying.